### DISCUSSION TITLE
Subnet Utilization View/Planner

### DISCUSSION THREAD

**Discussion by DCMattyG** [Azure/ipam#50] [Ideas]
Hi @letmetechyou, I'm moving this to a discussion as it's a great topic and something I've began to work on.

The ask from Issue #41 was to have some way of visualizing not only the currently utilized subnet blocks in an Azure vNET, but a tool to better understand what blocks are available with varying CIDR masks, similar to this example from the phpIPAM utility...

![197075837-4ccfbfed-0ab6-4fb9-bf44-343781422070](https://user-images.githubusercontent.com/12955732/200874433-4fd2ab35-12e6-43ff-a454-517558b97152.png)

**Comment by binnykmathew** [comment]
@DCMattyG Is there already a way in IPAM to find the next available CIDR range inside of a Vnet for a desired Subnet size ? 
Something like this does -> https://techcommunity.microsoft.com/t5/azure-networking-blog/programmatically-find-next-available-cidr-for-subnet/ba-p/3266016

**Reply by DCMattyG** [reply]
Hi @binnykmathew, there isn't yet, but you're the 3rd person that has requested this so I'm going to move it to the top of my list.

Would you be consuming this as an API, or would you also like it exposed in the IPAM UI? If you'd like to see it in the UI, please give me a few details as to how you picture that user flow working so I can use that to better design the experience.

Thanks!

**Reply by binnykmathew** [reply]
@DCMattyG Thanks for the quick reply.
Initially, we would like to get an API for this, so we can integrate with IaC and Internal Developer platforms.

**Reply by DCMattyG** [reply]
Wow @binnykmathew, this is AWESOME! Thank you for such a complete example and explanation!

I'll get to work right away on the API, and once we get that working, I'll work closely with you on a UI experience that makes sense for everyone.

**Reply by binnykmathew** [reply]
@DCMattyG Just wondering if there is a date when you will have the Subnet API ready for us to use in Production :)

**Reply by DCMattyG** [reply]
Hi @binnykmathew, we have a public holiday today in the US and I'm planning on using this time to try and get this into a Development image for you to test and provide feedback on. Hopefully you'll see an update from me within the next 1-2 days on this item.

**Comment by iamastranger** [comment]
@DCMattyG - I would vote for the following features, in the priority order
a) Subnet reservation using API - It will then make this tool feature complete :). Currently, this is a key feature thats missing for us to use this in an IaC automation.
b) Show in the portal all the reservations and done by which user
c) Performing a resevation via the portal is nice to have.
d) Subnet planner

Just wondering if there is any timeline when point (a) will be available ?

**Reply by DCMattyG** [reply]
Hi @iamastranger, thank you so much for the valuable feedback. A few notes/questions for you....

A) I have been working on an API that will provide you the next subnet of size X in a vNET. At the moment, I hadn't planned on making this a full-blown reservation like how the vNET reservations work, simply because there's far less chance of clash within a vNET to make subnets as opposed to multiple users carving out a Block for new vNET(s). Do you feel that this actually needs a complete reservation system in place, or would an API that can return the next subnet of a given mask size be sufficient?
B) This can already be done today. IPAM Admins can see **all** reservations, whereas non-admins can only see their own within the UI.
C) I've had multiple requests for this, and I'm happy to do it...but I keep thinking to myself "why?". If you are in the UI and can click a few buttons to manually make a reservation, why not just use those clicks to make the vNET in Azure instead? I'll create the functionality, but I keep wondering haha...
D) I've spent quite a bit of time on this over the holiday break. Hoping to have it complete by the new year.

For item A, I am hoping to have a functional test ready by the new year for feedback, but at this moment it will be an API that will reveal the next subnet of mask size X within a given vNET, and not a complete reservation experience. If we need the full reservation experience for subnets, please provide me a few details as to how this will help (example workflow where simply returning the next available subnet would not be sufficient) and of course I can work on getting it implemented.

Thank you for all of the valuable feedback, it's exactly what I wanted for Christmas ;-)

**Reply by iamastranger** [reply]
@DCMattyG Thankyou for a quick reply even during a holiday break :), shows your dedication towards making this project usefull.

- A) We just need an API that can return the next subnet of a given mask size. The use case is that we plan to integrate this with terraform and/or bicep at the time of subnet creation, so the returned CIDR is immediately used and doesnt require a reservation.
- B) Thanks, I was not aware of it. Thought, it was only possible to see those via a GET call to reservation API.
- C) In our case, we donot require this feature at all. We are happy with the API. 
- D) Cool 🥇.  Today, my team keeps track of this via a wiki page, so definitely a New year gift :)

Wish you a Happy New Year ;-)

**Comment by DCMattyG** [comment]
Alright @binnykmathew, @iamastranger, and @letmetechyou....finally got some things to show off here!

I'm just about ready to push some major updates, but I'd love some feedback from each of you.

In order to check out the changes, please switch the container images your App Service is pointing to, by appending "dev" to each of them as follows:

![image](https://user-images.githubusercontent.com/12955732/212999634-848e3250-30e2-46af-b041-99691353fcf7.png)

There are A LOT of changes, but the specific ones interesting to this thread are:

1) Subnet Planner (Available from the Tools > Planner menu):

![image](https://user-images.githubusercontent.com/12955732/212999831-672034ec-146c-44ec-a3a1-c7aa3e5d7a29.png)

> NOTE: The filter toggle in the upper right will allow you to flip between vNEt's mapped to a Space/Block, or **all** vNET's.

2) Experimental "Next Subnet from Virtual Network" API:

![image](https://user-images.githubusercontent.com/12955732/212999898-1007cf07-686f-4724-b5d9-a42e639de8fe.png)

POST to this endpoint with the following body:

```
{
    "vnet_id": <Full Resource ID of the Target Virtual Network (string)>,
    "size": <Network Mask bits for target Subnet (integer)>
}
```

> NOTE: I made this in the "tools" path because you can give it **ANY** vNET ID you have access to, not just one mapped to a space/block.

Looking forward to your feedback and getting this pushed out to live soon!

**Reply by binnykmathew** [reply]
@DCMattyG Thanks for making these features available.

We have tested this in our produciton tenant pointing to the dev images, below is our initial feedback.

The API works great when we use it in a green field i.e empty VNET and then use the API to reserve subnets. It starts picking from the top. However, it doesnt work as expected when we have a brown field, i.e if we already have some subnets that we have manually picked, lets say from the end. Then using this API, it starts to pick from the top breaking our largest CIDR ranges, which we donot want. 

See below an example:
Let say this is the current state for a VNET, the green box showing what we are already using.

![image](https://user-images.githubusercontent.com/26581533/213920929-98ed61cb-b632-4875-8f53-487547fc2b21.png)


Now, when we use the API it returns this

`{
    "cidr": "10.183.112.0/27"
}`

But, our expectation is it **shouldnt** divide a large CIDR block ( that we may want to use for other purposes ). It should have ideally divided the smallest CIDR ( which is 10.183.119.128/26 ) and returned either 10.183.119.128/27 or 10.183.119.160/27 


Regarding the planner, works really great . However, there is a minor cosmetic bug, where the "mask" field gives "undefined" error when I type something like "/27". It works fine, if I choose from the filter.

**Reply by DCMattyG** [reply]
Thanks so much for the feedback @binnykmathew! Funny I as just talking about something similar with @whizzer929 in #72.

So, if I add a parameter to the **reservations** API, something like `rangeFrom:` and the values could be `"start"` or `"end"`, would that solve your situation? Essentially telling it to find the closes CIDR starting from either the beginning or the end of the range?

Example:

```
{
    blocks: ['Block01', 'Block02'],
    size: 24,
    rangeFrom: "end"
}
```

Alternatively, I could do a **boolean** such as `fromEnd:` with values of `true` or `false` (defaulting to `false`)...

Example:

```
{
    blocks: ['Block01', 'Block02'],
    size: 24,
    fromEnd: true
}
```

Thoughts? Preference?

Also, I'll look into the *MASK* issue right away!

**Reply by whizzer929** [reply]
Since most use cases should be to pull from the beginning, I would have that as the default behavior.  Then have fromEnd: true as an optional  parameter.

**Reply by DCMattyG** [reply]
Thanks for the feedback @whizzer929! @binnykmathew, would this solve your situation as well, or were you looking for some more "advanced" logic in the CIDR selection?

The way it works now is that it tries to group all of the CIDR allocations as close together as possible, leaving large blocks at the end, and stuffing in smaller blocks where it can. With the flag I proposed above, we can reverse this logic, having it search from the end first. Anything "in the middle" would be more complex for sure, but doable.....just want some input from you first.

Much appreciated!

**Reply by DCMattyG** [reply]
BTW, the MASK issue in the Planner should be resolved if you restart your App Service to get the latest DEV images. Also, I added Dark Mode to the UI finally!

![image](https://user-images.githubusercontent.com/12955732/214764294-10d90d2f-d1fa-420d-b798-4eff772b52bb.png)

**Reply by binnykmathew** [reply]
> Thanks for the feedback @whizzer929! @binnykmathew, would this solve your situation as well, or were you looking for some more "advanced" logic in the CIDR selection?
> 
> The way it works now is that it tries to group all of the CIDR allocations as close together as possible, leaving large blocks at the end, and stuffing in smaller blocks where it can. With the flag I proposed above, we can reverse this logic, having it search from the end first. Anything "in the middle" would be more complex for sure, but doable.....just want some input from you first.
> 
> Much appreciated!

@DCMattyG Unfortunately it doesnt solve our problem ( or I guess majority of future users) by having a flag. Our intention of using the API is to integrate with our IaC where it automatically picks the next available CIDR of intended size from the smallest possible block and create the subnet. Here we cannot intervene and switch the flag as different VNets have unused blocks randomly in between ( for example, we may have decommissioned a service and deleted the associated subnet, freeing up that CIDR in the middle)

If we have to manually pick the Subnet, we could always use the Subnet planner. The API is for automation purpose and it should have the intelligence to return the correct size (from the smallest blocks) without breaking the larger blocks.

Thanks for fixing the Mask issue :)

**Reply by DCMattyG** [reply]
Fair enough @binnykmathew, would perhaps another flag that would find the smallest possible CIDR Block (within the target block) and carve the new Network from that....would that work for you?

**Reply by binnykmathew** [reply]
@DCMattyG Yes, that should work :)

Thanks

**Reply by DCMattyG** [reply]
Ok, based on the above discussion, let me propose this...

The new flag would be something like `method` or `allocationMethod` and would have several options such as:

**Find next CIDR, starting from the *BEGINNING* of the Block(s) (DEFAULT):**
```
{
    blocks: ['Block01', 'Block02'],
    size: 24,
    method: "fromStart"
}
```

**Find next CIDR, starting from the *END* of the Block(s):**
```
{
    blocks: ['Block01', 'Block02'],
    size: 24,
    method: "fromEnd"
}
```

**Find next CIDR, using the *SMALLEST POSSIBLE CIDR* available in the Block(s):**
```
{
    blocks: ['Block01', 'Block02'],
    size: 24,
    method: "fromSmallest"
}
```

Any thoughts/improvements/opposition to this mechanism? (@binnykmathew & @whizzer929)

**Reply by binnykmathew** [reply]
@DCMattyG Can you clarify what is this "blocks" here ?
Your original API required a VNet ID, so is this Block in addition to that , if so is it optional ?

**Reply by DCMattyG** [reply]
**Blocks** is a list of Block *names* that exist within the **Space**, in this context.

**Reply by DCMattyG** [reply]
@binnykmathew, I think I'm mixing up two things here....

The directly above section is talking about creating a reservation from a Block in IPAM...either at the block level directly, or at the Space level by way of providing a list of Block from which to search for an available CIDR.

I think what you're referring to is not the reservations per-se, but the "Next subnet available in vNET" API....correct? If that is the case, it would look something like this:

```
{
    "vnet_id": <String: Full Resource ID of the Target Virtual Network>,
    "size": <Integer: Network Mask bits for target Subnet>,
    "reverse_search": <Boolean: If true, new subnets will be selected from as close to the END of the vNET as possible>,
    "smallest_cidr": <Boolean: If true, new subnets will be selected from the smallest available CIDR within the target vNET>
}
```

In this case, you can use **reverse_search**, **smallest_cidr**, or **both** depending on your use case.

Hope that clears things up!

**Reply by binnykmathew** [reply]
@DCMattyG Thanks for the clarification. Sorry that some of my replies also contributed to this confusion. All my replies in this thread has been in context of picking a Subnet from a Vnet. Yes, the above proposed changes (smallest_cidr reverse_search) in the "Next subnet available in vNET" API looks great to me. 

_(Edited )_
Regarding the Reservation API, I dont have strong opinion on that. We are already using this API and as long as the "Blocks" field is not mandatory in the body , Iam ok with it. The changes that you make should be backward compatible.

**Reply by DCMattyG** [reply]
Good morning @binnykmathew. Completely understood, and allow me to provide some clarity. The above is a mix where we've been talking about 3 different API's.

The first one, which is the one you're referring to, is as follows:

![image](https://user-images.githubusercontent.com/12955732/216411293-84c88d9d-62a0-42a2-8099-9c8477c33518.png)

You can see this API lives at the **Block** level, and as such doesn't require a list of Blocks as it uses the Block which is a part of the API url. Here we've added the **reverse_search** and **smallest_cidr** flags, however those are optional so the API is indeed backwards compatible.

The next one comes from a use case where you want to select a new CIDR from more thank one Block. It's format is as follows:

![image](https://user-images.githubusercontent.com/12955732/216412380-9ed2a8c7-172a-444d-938a-48c2865aa948.png)

You can see here, since this API lives at the **Space** level, it requires a list of one or more Blocks that live beneath that Space from which to pull a new CIDR from. This could be a case where you have non-contiguous IP space, which may live on the same Azure Virtual Network, but be a part of two different Blocks as it cannot be summarized into a single CIDR. The longer term solution here is to allow more than one CIDR for a Block, but this is a simple way to address this without changing the architecture. This also has the **reverse_search** and **smallest_cidr** flags.

Finally, there is the **nextAvailableSubnet** API, which is designed as follows:

![image](https://user-images.githubusercontent.com/12955732/216414176-4e3f68e1-60a6-48c3-bae9-502cb9ba81c8.png)

I've incorporated the same **reverse_search** and **smallest_cidr** flags here as well so the behavior is consistent across all 3 API's.

I hope that clears things up, you can test out the new DEV code by following the directions [HERE](https://github.com/Azure/ipam/discussions/50#discussioncomment-4710593) (also can be found just above).

Please let me know if you need any further clarification, or if you have any issues with the DEV code.

Thank you as always for helping to make Azure IPAM better for everyone!

**Reply by binnykmathew** [reply]
@DCMattyG Thanks a lot for the clarification.
We will test it this week and let you know.
